compiler/hash-utils/src/temp_writer.rs

```diff
@@ -0,0 +1,24 @@
+//! Utility for temporarily writing something to a buffer.
+use std::io;
+
+use derive_more::Constructor;
+
+#[derive(Constructor, Default)]
+pub struct TempWriter(Vec<u8>);
+
+impl TempWriter {
```

Comment:
Why not just use `BufWriter`?

Reply:
Where is that from?

Reply:
It feels like the [`BufWritter`](https://doc.rust-lang.org/std/io/struct.BufWriter.html#:~:text=For%20example%2C%20every%20call%20to,same%20file%20or%20network%20socket.) API is a bit more awkward to use than this.

Reply:
Okay fair. Why not just use a String and preallocate capacity?

Reply:
From the looks of it, [Write](https://doc.rust-lang.org/std/io/trait.Write.html) is implemented for a few types (one of them being `Vec<u8>`, which means we are limited to a number of in-built implementations (foreign implentations restriction) it means we can't simply just use a `String` as a buffer or we would still have to wrap it in a `TempWriter`.

Reply:
The alternative is to directly bake an `str` type and then represent it as an `Alloc`, i.e. `<size> <ptr_to_another_alloc>` and then the `ptr_to_another_alloc` is the underlying bytes buffer for the string.

Reply:
it does not implement Write, but why not use String for the purposes of TempWriter. As I understand it it is only used for strings at the end of the day?

Reply:
I mean a `String` and `Vec<u8>` are the same under the hood, so I think it doesn't matter too much.